source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,487 @@
+SubsetLoadBalancer::SubsetPtr SubsetLoadBalancer::findSubset(
+    const std::vector<Router::MetadataMatchCriterionConstSharedPtr>& matches) {
+  LbSubsetMap* subsets = &subsets_;
+  for (size_t i = 0; i < matches.size(); i++) {
+    const Router::MetadataMatchCriterionConstSharedPtr& match = matches[i];
+    const auto& subset_it = subsets->find(match->name());
+    if (subset_it == subsets->end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    const ValueSubsetMap& vs_map = subset_it->second;
+    const auto& vs_it = vs_map.find(match->value());
+    if (vs_it == vs_map.end()) {
+      // No subsets with this value.
+      break;
+    }
+
+    const LbSubsetEntryPtr& entry = vs_it->second;
+    if (entry->host_subset_ != nullptr && i + 1 == matches.size()) {
+      return entry->host_subset_;
+    }
+
+    subsets = &entry->children_;
+  }
+
+  return nullptr;
+}
+
+// Given the addition and/or removal of hosts, update all subsets to
+// match, creating new subsets as necessary.
+void SubsetLoadBalancer::update(const std::vector<HostSharedPtr>& hosts_added,
+                                const std::vector<HostSharedPtr>& hosts_removed, bool local) {
+  RELEASE_ASSERT(!local || original_local_host_set_);
+
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) {
+    if (default_host_subset_ == nullptr) {
+      // First update: create the default host subset with all hosts
+      default_host_subset_ = newSubset([](const HostSharedPtr&) -> bool { return true; });
+      default_host_subset_->lb_.reset(newLoadBalancer(default_host_subset_));
+    } else {
+      if (local) {
+        default_host_subset_->local_host_set_.update(hosts_added, hosts_removed);
+      } else {
+        default_host_subset_->host_set_.update(hosts_added, hosts_removed);
+      }
+    }
+  } else if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::DEFAULT_SUBSET) {
+    if (default_host_subset_ == nullptr) {
+      // First update: create the default host subset
+      default_host_subset_ = newSubset(
+          std::bind(&SubsetLoadBalancer::hostMatchesDefaultSubset, this, std::placeholders::_1));
+      default_host_subset_->lb_.reset(newLoadBalancer(default_host_subset_));
+    } else {
+      // Filter out added hosts that don't match the default subset
+      std::vector<HostSharedPtr> subset_hosts_added;
+      for (const auto& host : hosts_added) {
+        if (hostMatchesDefaultSubset(host)) {
+          subset_hosts_added.emplace_back(host);
+        }
+      }
+
+      if (!subset_hosts_added.empty() || !hosts_removed.empty()) {
+        if (local) {
+          default_host_subset_->local_host_set_.update(subset_hosts_added, hosts_removed);
+        } else {
+          default_host_subset_->host_set_.update(subset_hosts_added, hosts_removed);
+        }
+      }
+    }
+  }
+
+  std::unordered_set<SubsetPtr> subsets_created;
```

Comment:
This method is fairly complex internally, I think it could do with some comments guiding the reader through what's going on at least at each major block. Locally, the code looks good, but it would be good to have the intuition conveyed about what is happening.

Reply:
✅